### DISCUSSION TITLE
Struggling with the converting time and understanding --help

### DISCUSSION THREAD

**Discussion by Mannshoch** [IfcOpenShell/IfcOpenShell#2184] [Q&A]
Im struggling a bit with the --help output. 
What does the --model argument mean? I do not understand the difference that it made.
```
                                        Specifies whether to include surfaces
                                        and solids in the output result.
                                        Typically these are representations of
                                        type Body or Facetation. Included by
                                        default.
```
Other Question. Currently Converting a 300MB IFC needs smuch time. The first part is fast but getting from `Done creating geometry (23837 objects)` to the final file even needs hours even  `-j 10` is active. Are threre any settings to speedup the process?
I need The bodies from the Quality of an "Extrusion" not less if possible.

**Comment by Mannshoch** [comment]
My last test with converting was able to go from Start ( `-j 10` ) to `"Done Creating geometry"` in about 4min.
```
IfcOpenShell IfcConvert v0.7.0-1b1fd1e6 (OCC 7.5.3)
Scanning file...
Done scanning file
Parsing input file took 28 seconds
Creating geometry...
Done creating geometry (23837 objects)
```
But here it stuck since 13h with a permanent CPU usage of 10%

**Reply by Mannshoch** [reply]
Ther are cases on huge files where fcConvert simply stop calculating wthout any error on the cli.

**Reply by Moult** [reply]
Unfortunately there isn't really an easy answer, as it could be due to various reasons. There are some documented here (yes it's for the BlenderBIM Add-on, but some apply to non-Blender generic geometry processing too:

 - https://blenderbim.org/docs/users/dealing_with_large_models.html#large-filesizes
 - https://blenderbim.org/docs/users/dealing_with_large_models.html#model-breakdown-strategies
 - https://blenderbim.org/docs/users/dealing_with_large_models.html#large-polygon-counts
 - https://blenderbim.org/docs/users/dealing_with_large_models.html#excessive-or-low-quality-booleans

You may try using the BlenderBIM Add-on to do a visual inspection, and try various filtered processing.

**Reply by Mannshoch** [reply]
Interesting, thanks. Unfortunately I have to work with what I get. Most Architect are not able to reduce file-size to the Rooms I need. Normally I get the whole Building with all details. Discussion is mostly meaningless. 

I'm checking If I could work with `--include`, but that means I need an IFC viewer and analyses the model instead doing it in my preferred CAD environment.

The most time of the converting is needed after `Done Creating geometry (7468 objects)` also this part after that seems is calculated on only one core. What is happen there? There is no progression information on cli. On Task Manager you could see the RAM usage increasing to ridiculous heights.

**Reply by Moult** [reply]
Have you considered using the BlenderBIM Add-on? It has quite a few ways of customising (including limiting loading content to just rooms you need) the IFC loading. Afterwards you can just save it to an obj/gtlf/whatever format you were using in IfcConvert (with a few exceptions, like step).

Stuck on one core could suggest it's stuck on a few (or even one) problematic objects. RAM could be either large poly counts or funky booleans or a combination.

**Reply by Mannshoch** [reply]
My first goal was to have a script where I could put all customer IFC and automatically get STEP files. This way I'm not time limitted and could work on other thing. The moment I have to check a file detect the correct data and feed the converting, I have to use time for something, brake some time and work again in an other tool and check if I really have all needed data.

I'm not a developer bur If you give advice I follow and may could help detect some problems or possibilites for enhances.

**Reply by Moult** [reply]
Unfortunately if a file has quirky geometry there aren't that many options other than pausing and checking manually.

**Reply by Mannshoch** [reply]
It's a building I don't know how quirky that could be. OK, there are architects with strange Ideas.

I did a test and Updated Windows rebooted and closed all not used application and let them run the whole day. Now the conversion worked. It seems it had 281minutes for the work.

**Reply by Mannshoch** [reply]
Do you may know how diferent the way IFC saves a body is compared to STEP 242? 
There are rumors that IFC use STEP 242 inside to save bodies. With that in mind may STEP 242 could be faster.

**Reply by Moult** [reply]
Quirky geometry is not typically intentional. Most users create IFCs through exporting, of which the behaviour is determined by a myriad (or even worse, none at all) of export settings. A common example is when a solid modeler (e.g. Revit) exports solids as a BRep and performs its own tessellation that the user cannot control. Here's a simple example where [a basic shape multiplies by filesize by 27X](https://wiki.osarch.org/index.php?title=Revit_setup_for_OpenBIM/Revit_and_IFC_Geometry#Reducing_curves_in_profiles_that_are_swept.2C_revolved.2C_or_blended) and any user who doesn't know this quirk of Revit will probably have this issue. There's not much we can do except raise awareness so that vendors can improve their IFC support, promote an ecosystem of NativeIFC development, and also educate users on how to avoid these pitfalls, but the industry is still very young in this regard.

IFC borrows a lot from STEP, but nothing new from STEP 242 from my understanding that would lead to any speed increase. Also, IfcConvert works generically through an IFC --(tessellate)--> OpenCascade --(serialize)--> OBJ/DAE/glTF/STP/etc although in theory going to STEP-based things (of which STEP 242 is not special) could be faster by bypassing OpenCascade and just converting the underlying implicit IFC keywords (though this is has risks in itself), it's just not how IfcConvert works.

**Reply by Mannshoch** [reply]
As far I know "swept solid" is best Quality of data, than Brep and Tesselation is lowest Quality of a body.
(https://www.cadinterop.com/en/your-needs/step/different-geometry-types-and-formats-in-cad.html)

Do I understand it right, that tesselate means that OpenCascade create kind of a low Quality body and after that It tries to calculate again a Brep body? 
Swept Solid -> Tesselated Body -> Brep

**Reply by Moult** [reply]
There's no such thing as best quality of data, it depends on what you're doing. A swept solid might be best quality of data for a standard extrusion. However, a brep or tessellation is the highest quality of data for a scan, as-built, dilapidation model, or mesh where UV is important for textures. Nurbs are best if you're doing a nurby design, and a triangulated faceset is best for when you need a TIN.

So Revit only works with solids, and depending on the version of Revit, may tessellate all the solids and produce breps/tessellations. In recent versions, it goes to the other extreme, making everything as an AdvancedBrep which is overkill. This is nothing to do with OpenCascade, this is what Revit does in its export process before IfcOpenShell and OpenCascade gets involved.

(Of course, the IFC may not come from Revit, so I'm just speculating here)

**Reply by aothms** [reply]
Just a quick not that with

`IfcConvert -vvvv` (4 v-s but 2 also already works)

you get detailed and realtime logging on what happens. That'll help you understand what particular element is taking so long.

**Reply by Mannshoch** [reply]
I fully agree with your definition of Quality but mine understanding was not driven by the Application which use data. It's defined by the lose of information. If you convert data into tesselated data and convert it back you e.g. lose the possibility to select a filet and get the correct value. I really would prefer a loosless coverting of geometric data... or at least as loosles as possible.

**Reply by Moult** [reply]
@Mannshoch yes, if you started with an extrusion then ended up with a meshy thing but needed information from the extrusion (e.g. original profile, any arcs) then you've lost data. The original issue was with quirky geometry, and a lot of IFC exporters tend to auto-convert solid geometry to meshy things, sometimes very poorly and mostly without user intervention, leading to quirky geometry.
